content/05_Filtering.md

```diff
@@ -1,269 +1,314 @@
+
+`Single` and `SingleAsync` share with `Last` and `LastAsync` the characteristic that they don't initially know when they receive an item from the source whether it should be the output. That may seem odd: since `Single` requires the source stream to provide just one item, surely it must know that the item it will deliver to its subscriber will be the first item it receives. This is true, but the thing it doesn't yet know when it receives the first item is whether the source is going to produce a second one. It can't forward the first item unless and until the source completes. We could say that `SingleAsync`'s job is to first verify that the source contains exactly one item, and then to forward that item if it does, but to report an error if it does not. In the error case, `SingleAsync` will know it has gone wrong if it ever receives a second item, so it can immediately call `OnError` on its subscriber at that point. But in the success scenario, it can't know that all is well until the source confirms that nothing more is coming by completing. Only then will `SingleAsync` emit the result.
+
+### Skip and SkipLast
+
+What if we want the exact opposite of the `Take` or `TakeLast` operators? Instead of taking the first 5 items from a source, maybe I want to discard the first 5 items instead? Perhaps I have some `IObservable<float>` taking readings from a sensor, and I have discovered that the sensor produces garbage values for its first few readings, so I'd like to ignore those, and only start listening once it has settled down. I can achieve this with `Skip(5)`.
 
-## Skip and Take						
+`SkipLast` does the same thing at the end of the sequence—it omits the specified number of elements at the tail end. As with some of the other operators we've just been looking at, this has to deal with the problem that it can't tell when it's near the end of the sequence. It only gets to discover which were the last (say) 4 elements after the source has emitted all of them, followed by an `OnComplete`. So `SkipLast` will introduce a delay. If you use `SkipLast(4)`, it won't forward the first element that the source produces until the source produces a 5th element. So it doesn't need to wait for `OnCompleted` or `OnError` before it can start doing things, it just has to wait until its certain that an element is not one of the ones we want to discard.
 
 The other key methods to filtering are so similar I think we can look at them as one big group. First we will look at `Skip` and `Take`. These act just like they do for the `IEnumerable<T>` implementations. These are the most simple and probably the most used of the Skip/Take methods. Both methods just have the one parameter; the number of values to skip or to take.
```

Comment:
This feels like this has all got a bit messed up in the edit. Is this paragraph still needed?

Reply:
Given how long it took me to spot that the odd/even thing was actually a comment about the variable names it's possible that I've got brain fog this morning... But I'm not seeing the issue here.

I think this whole section was out of order - the **Skip and SkipLast** section it now belongs to should have followed on from **TakeLast**, and the fact that the **SingleAsync and SingleOrDefaultAsync** section was in between those looks like an aberration.

But having fixed that ordering, I think this paragraph is still required. It's the only paragraph that describes what `SkipLast` does. And although the _end of sequence_ problems it has to deal with are similar to `Last` and `TakeLast` they're not the same. (`Last` and `TakeLast` can't give you _anything_ until they've seen the end of the sequence. But `SkipLast` can start giving you items earlier. If you've asked to skip the last 4 items, and it has now received 5 items, it knows for a fact that it won't be skipping the 1st item, so it can give it to you without waiting for the source to complete.)

So although there are similarities to the issues dealt with by other operators, there are also differences, so I think the explanation of that aspect in this paragraph also remains.

So I'm not sure there's anything here that isn't needed. What have I missed?

Reply:
I think I got lost in the diff.